doc/api/async_hooks.md

```diff
@@ -0,0 +1,671 @@
+`init()` to fire. Generally it will be the name of the resource's constructor.
+Some examples include `TCP`, `GetAddrInfo` and `HTTPParser`. Users will be able
+to define their own `type` when using the public embedder API.
+
+**Note:** It's possible to have name collisions. So embedders are recommended
+to use a prefix of sorts to prevent this.
+
+`triggerId` is the `id` of the resource that caused (or "triggered") the
+new resource to initialize and that caused `init()` to fire.
+
+The following is a simple demonstration of this:
+
+```js
+const async_hooks = require('async_hooks');
+
+asyns_hooks.createHook({
+  init: (id, type, triggerId) => {
+    const cId = async_hooks.currentId();
+    process._rawDebug(`${type}(${id}): trigger: ${triggerId} scope: ${cId}`);
+  }
+}).enable();
+
+require('net').createServer(c => {}).listen(8080);
+```
+
+Output when hitting the server with `nc localhost 8080`:
+
+```
+TCPWRAP(2): trigger: 1 scope: 1
+TCPWRAP(4): trigger: 2 scope: 0
+```
+
+The second `TCPWRAP` is the new connection from the client. When a new
+connection is made the `TCPWrap` instance is immediately constructed. This
+happens outside of any JavaScript stack (side note: a `currentId()` of `0`
+means it's being executed in "the void", with no JavaScript stack above it).
+With only that information it would be impossible to link resources together in
+terms of what caused them to be created. So `triggerId` is given the task of
+propagating what resource is responsible for the new resource's existence.
+
+Below is another example with additional information about the calls to
+`init()` between the `before()` and `after()` calls. Specifically what the
+callback to `listen()` will look like. The output formatting is slightly more
+elaborate to make calling context easier to see.
+
+```js
+const async_hooks = require('async_hooks');
+
+let ws = 0;
+async_hooks.createHook({
+  init: (id, type, triggerId) => {
+    const cId = async_hooks.currentId();
+    process._rawDebug(' '.repeat(ws) +
+                      `${type}(${id}): trigger: ${triggerId} scope: ${cId}`);
+  },
+  before: (id) => {
+    process._rawDebug(' '.repeat(ws) + 'before: ', id);
+    ws += 2;
+  },
+  after: (id) => {
+    ws -= 2;
+    process._rawDebug(' '.repeat(ws) + 'after:  ', id);
+  },
+  destroy: (id) => {
+    process._rawDebug(' '.repeat(ws) + 'destroy:', id);
+  },
+}).enable();
+
+require('net').createServer(() => {}).listen(8080, () => {
+  // Let's wait 10ms before logging the server started.
+  setTimeout(() => {
+    console.log('>>>', async_hooks.currentId());
+  }, 10);
+});
+```
+
+Output from only starting the server:
+
+```
+TCPWRAP(2): trigger: 1 scope: 1
+# .listen()
+TickObject(3): trigger: 1 scope: 1
+before:  3
+  Timeout(4): trigger: 3 scope: 3
+  TIMERWRAP(5): trigger: 3 scope: 3
+after:   3
+before:  5
+  before:  4
+    TTYWRAP(6): trigger: 4 scope: 4
+    SIGNALWRAP(7): trigger: 4 scope: 4
+>>> 4
+  after:   4
+after:   5
```

Comment:
Updated the output .. didn't see any changes to the script.
Note that we're using class like function declarations (instead of arrow functions) as per request by @Fishrock123